pkg/controller/etcdproxy/certs.go

```diff
@@ -110,146 +118,142 @@ func (c *EtcdProxyController) setNewAPIServerCertificates(etcdstorage *etcdstora
 	return utilerrors.NewAggregate(errs)
 }
 
-// createEtcdProxyClientCAConfigMap creates ConfigMap in controller namespace with Etcd Proxy CA bundle
-// for verifying incoming client certificates.
-func (c EtcdProxyController) createEtcdProxyClientCAConfigMap(etcdstorage *etcdstoragev1alpha1.EtcdStorage,
-	clientSingerCert []byte) error {
-	// ConfigMap in controller namespace for the etcd proxy CA certificate.
-	_, err := c.kubeclientset.CoreV1().ConfigMaps(c.config.ControllerNamespace).
-		Get(etcdProxyCAConfigMapName(etcdstorage), metav1.GetOptions{})
+// ensureCAConfigMap ensures that ConfigMap containing the CA bundle is created and has correct CA bundle.
+func (c *EtcdProxyController) ensureCAConfigMap(etcdstorage *etcdstoragev1alpha1.EtcdStorage,
+	configMapNamespace, configMapName, caFileName string, caBytes []byte, expiryDate time.Time) error {
+	caConfigMap, err := c.kubeclientset.CoreV1().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
 	if errors.IsNotFound(err) {
-		data := map[string]string{"client-ca.crt": string(clientSingerCert)}
-		_, err = c.kubeclientset.CoreV1().ConfigMaps(c.config.ControllerNamespace).
-			Create(newConfigMap(etcdstorage, etcdProxyCAConfigMapName(etcdstorage), c.config.ControllerNamespace, data))
+		// ConfigMap is not found, so we're creating a new one.
+		// This can fail when creating a ConfigMap in the API server namespace if there are no appropriate RBAC rule.
+		data := map[string]string{
+			caFileName: string(caBytes),
+		}
+		// TODO: Annotations will not work for CA ConfigMaps as they can have mulitple certs in bundle. This is required so we have a check are certs generated until we move to new implementation.
+		annotations := map[string]string{
+			ProxyCertificateExpiryAnnotation: expiryDate.Format(time.RFC3339),
+		}
+
+		caConfigMap, err = c.kubeclientset.CoreV1().ConfigMaps(configMapNamespace).Create(newConfigMap(etcdstorage, configMapName, configMapNamespace, annotations, data))
 		if err != nil {
-			// TODO: refactor event handling (hint: see #40).
-			c.recorder.Event(etcdstorage, corev1.EventTypeWarning, ErrUnknown, err.Error())
 			return err
 		}
+
+		// If we're in controller namespace, put appropriate OwnerRef.
+		if configMapNamespace == c.config.ControllerNamespace {
```

Comment:
This smells bad.  It will be less objectionable if you check if the namespaces between etcdstorageand configmap are the same.